src/sgame/sg_buildable.cpp

```diff
@@ -792,39 +795,57 @@ void G_UpdateBuildablePowerStates()
 			// TODO: Refer to a SpawnerComponent here.
 			if (entity.Get<TelenodeComponent>() || entity.Get<EggComponent>()) return;
 
-			// Power all other buildables depending on main buildable state for now.
-			if (activeMainBuildable) {
-				buildableComponent.SetPowerState(true);
-			} else {
+			// Power off all buildables if there is no main buildable.
+			if (!activeMainBuildable) {
 				buildableComponent.SetPowerState(false);
 				return;
 			}
 
 			// In order to make good a deficit, don't shut down buildables that have no cost.
 			if (BG_Buildable(entity.oldEnt->s.modelindex)->buildPoints <= 0) return;
-
-			buildables.push_back(&entity);
+			if (!entity.oldEnt->powered) {
+				unpoweredBuildables.push_back(&entity);
+				unpoweredBuildableTotal += BG_Buildable(entity.oldEnt->s.modelindex)->buildPoints;
+			} else {
+				buildables.push_back(&entity);
+			}
 		});
 
 		// If there is no active main buildable, all buildables that can shut down already did so.
 		if (!activeMainBuildable) continue;
 
-		int deficit = level.team[team].spentBudget - (int)level.team[team].totalBudget;
+		// Positive deficit means that we are over, and negative means we have a surplus.
+		int deficit = level.team[team].spentBudget - (int)level.team[team].totalBudget - unpoweredBuildableTotal;
 
-		// Do not power down more buildables if there is no deficit.
-		if (deficit <= 0) continue;
+		// Exactly at our limit. Nothing else to do.
+		if (deficit == 0) continue;
 
-		buildables.sort(CompareBuildablesForPowerSaving);
+		// We have surplus bp, but nothing else to power on, so we're done here.
+		if (deficit < 0 && unpoweredBuildables.empty()) continue;
```

Comment:
I think you could remove this `if` block and the one before it.

Reply:
What do you mean? I think that the explicit cases are much clearer about the code intention.

Reply:
So `deficit < 0` and you want to power up as many of the N unpowered buildables as possible. I don't see anything qualitatively different about what happens when N ==0 from N != 0, so for me it is harder to read if there is a special case for N == 0.

Reply:
While its negligible, you'll also be unnecessarily going through a list of buildables. I think I'll leave this one as is.